src/Utility/Test/TestSplineUtils.C

```diff
@@ -450,3 +451,75 @@ TEST(TestSplineUtils, BuildKnotVector)
   for (size_t i = 0; i < ref.size(); ++i)
     EXPECT_FLOAT_EQ(ref[i], res[i]);
 }
+
+
+namespace {
+
+struct AdjustBasisTest {
+  int order_in;
+  std::vector<double> knots_in;
+  SplineUtils::AdjustOp op;
+  std::vector<double> knots_out;
+};
+
+using Op = SplineUtils::AdjustOp;
+
+const std::vector<AdjustBasisTest> tests = {
+  // p1 -> p1
+  {2, {0.0, 0.0, 1.0, 1.0}, Op::Original,
+      {0.0, 0.0, 1.0, 1.0}},
+
+  // p1 -> p2
+  {2, {0.0, 0.0, 1.0, 1.0}, Op::Raise,
+      {0.0, 0.0, 0.0, 1.0, 1.0, 1.0}},
+  // p1 -> p2, extra knot
+  {2, {0.0, 0.0, 0.5, 1.0, 1.0}, Op::Raise,
+      {0.0, 0.0, 0.0, 0.5, 1.0, 1.0, 1.0}},
+  // p2 -> p3, extra knot
+  {3, {0.0, 0.0, 0.0, 0.5, 1.0, 1.0, 1.0}, Op::Raise,
+      {0.0, 0.0, 0.0, 0.0, 0.5, 1.0, 1.0, 1.0, 1.0}},
+  // p2 -> p3, reduced cont
+  {3, {0.0, 0.0, 0.0, 0.5, 0.5, 1.0, 1.0, 1.0}, Op::Raise,
+      {0.0, 0.0, 0.0, 0.0, 0.5, 0.5, 0.5, 1.0, 1.0, 1.0, 1.0}},
+  //p3 -> p4
+  {4, {0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0}, Op::Raise,
+      {0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 1.0}},
+  //p3 -> p4, reduced cont
+  {4, {0.0, 0.0, 0.0, 0.0, 0.5, 0.5, 1.0, 1.0, 1.0, 1.0}, Op::Raise,
+      {0.0, 0.0, 0.0, 0.0, 0.0, 0.5, 0.5, 0.5, 1.0, 1.0, 1.0, 1.0, 1.0}},
+
+  //p2 -> p1
+  {3, {0.0, 0.0, 0.0, 1.0, 1.0, 1.0}, Op::Lower,
+      {0.0, 0.0, 1.0, 1.0}},
+  //p2 -> p1, extra knot
+  {3, {0.0, 0.0, 0.0, 0.5, 1.0, 1.0, 1.0}, Op::Lower,
+      {0.0, 0.0, 0.5, 1.0, 1.0}},
+  //p3 -> p2, reduced cont
+  {4, {0.0, 0.0, 0.0, 0.0, 0.5, 0.5, 1.0, 1.0, 1.0, 1.0}, Op::Lower,
+      {0.0, 0.0, 0.0, 0.5, 0.5, 1.0, 1.0, 1.0}},
+  //p4 -> p3
+  {5, {0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 1.0}, Op::Lower,
+      {0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0}},
+};
+
+}
+
+class TestSplineUtils : public testing::Test,
+                        public testing::WithParamInterface<AdjustBasisTest>
+{
+};
+
+
+TEST_P(TestSplineUtils, AdjustBasis)
+{
+  Go::BsplineBasis basis_in(GetParam().order_in,
+                            GetParam().knots_in.begin(),
+                            GetParam().knots_in.end());
+
+  const Go::BsplineBasis res = SplineUtils::adjustBasis(basis_in,GetParam().op);
+  const std::vector<double> out{res.begin(), res.end()};
```

Comment:
This looks suspicious, `begin()` and `end()` returns iterators, afaik, and then you put them into a double array? Or it this another way of copying the array.

Reply:
ok, I was confused by the syntax, I thought that constructor was invoked with parentheses, i.e. `out(res.begin(), res.end())`, and when using braces `{}` means a list of elements instead and here this would be a length-two vector with that start and stop iterator values as elements. But clearly that was not the case then..

Reply:
right. it's c++-11 uniform initialization. you can initialize arrays or classes or anything really using {} . as it has advantages, such as not allowing implicit conversions of arguments, i've made it a habit to use it (see https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Res-list for all the itty-gritty details if you are interested).